Add differential privacy as a threat mitigation (#91)
abetterinternet/prio-documents: draft-pda-protocol.md

1. The input validation protocol executed by the aggregators prevents either
individual clients or coalitions of clients from compromising the robustness
property.
1. If aggregator output satisifes differential privacy {{dp}}, then all records

> not leaked by malicious clients are still protected.

Reviewer: Would you mind adding a reference for `{{dp}}` and verify the document builds by running `make`? See the top of the document for guidance. If you don't have time, then please at least provide a link to an article you think would be a suitable reference. The most useful reference would be something that describes the distinction between client-side and server-side noise addition.

Reviewer: Oops, I missed that {{dp}} refers to a section and not a paper. Disregard.

   aggregations, joint randomness, arithmetic circuits).
1. Collectors may trivially defeat availability by discarding output shares
   submitted by aggregators.
1. Known input injection. Collectors may collude with clients to send known
   input to the aggregators, allowing collectors to shrink the effective

>    anonymity set by subtracting the known inputs from the final output.

Reviewer: I would work in the term "Sybil attack" here. While we're at it, let's add a reference for Sybil attacks (i.e., `{{sybil}}`). I think the canonical reference is https://link.springer.com/chapter/10.1007/3-540-45748-8_24.

Author: Done

the minimum batch size is too small. This document does not specify how to
choose minimum batch sizes.

> ## Differential privacy {#dp}

Reviewer 2: Besides this informal recommendation, do we need explicit protocol support for differential privacy so that collectors can de-noise outputs? We can leave it up to aggregators to decide how they're going to implement DP but I wonder if `PDAOutputShare` should have a field for the epsilon value that was used by the aggregator. Forgive me if I'm talking nonsense about DP, I am speaking in the terms that we used in Prio v2.

Author: Hm. I am nervous about being to prescriptive here. In the simplest protocol design nothing is needed since the epsilon is hardcoded into the specific protocol instantiation and won't change.

In practice, some specific instantiations may want to reveal even more information about how noise was applied, e.g.
- The distribution noise is sampled from (Laplace, Gaussian, etc)
- Parameters of the noise distribution
- Any kind of threshold used (for example, if you are using approximate DP)

I think we should make this as opaque to the protocol as possible vs. prescribing some single "epsilon" field which might be too constraining. What do you think?

Author: However, I still think this current PR is land-able given that a basic instantiation can hardcode everything without requiring any communication.

Reviewer: The way to go here, I think, is to document the open question by adding an `[OPEN ISSUE: blah blah blah]`.

Author: Done

Reviewer 2: I put a note about this in #19, which I think is an appropriate issue to track this discussion.

Optionally, PDA deployments can choose to ensure their output F achieves

> [differential privacy](https://en.wikipedia.org/wiki/Differential_privacy).

Reviewer: Given that the likely trajectory of this document is an IETF WG draft, it would be better to replace this markdown with a `{{dp}}` reference.

Reviewer: Edited this comment.

Author: Done, I added a general DP reference in place of the wikipedia link. It discusses central, local, and multi-party DP.

> A simple approach would require both the helper and leader to add two-sided

Reviewer: ```suggestion
A simple approach would require the aggregators to add two-sided
```
Depending on what we do for #68 there may be more than one helper.

noise (e.g. sampled from a two-sided geometric distribution) to outputs.
Since each aggregator is adding noise independently, privacy can be guaranteed

> even if one of the aggregators is malicious. Differential privacy is a strong

Reviewer: ```suggestion
even if all but one of the aggregators is malicious. Differential privacy is a strong
```
